File: scripts/address_validation_signer/ts/validation_signer.ts
/**
 * Reference SNIP-12 signer for the privacy contract's depositor validation.
 *
 * Produces a signature that `privacy::snip12::verify_depositor_validation`
 * accepts. The typed-data layout here MUST stay in lockstep with the Cairo
 * verifier and with the sibling `py/validation_signer.py` — any drift on
 * one side breaks all three.
 *
 * Usage:
 *   npm run sign -- \
 *     --signer-private-key 0x... \
 *     --depositor 0x... \
 *     --issued-at 1700000000 \
 *     --chain-id SN_SEPOLIA
 *
 * Or import { signDepositorValidation } from another script/test.
 */

import { ec, num, typedData, type TypedData } from "starknet";

export interface SignDepositorValidationInput {
  signerPrivateKey: string;
  depositor: string;
  issuedAt: number | bigint | string;
  chainId: string;
}

export interface SignDepositorValidationOutput {
  messageHash: string;
  signature: { r: string; s: string };
  signerPublicKey: string;
  input: {
    depositor: string;
    issuedAt: string;
    chainId: string;
  };
}

const SNIP12_DOMAIN_NAME = "Screening";
const SNIP12_DOMAIN_VERSION = "2";
const PRIMARY_TYPE = "DepositorValidation";

function buildTypedData(depositor: string, issuedAt: string, chainId: string): TypedData {
  return {
    domain: {
      name: SNIP12_DOMAIN_NAME,
      version: SNIP12_DOMAIN_VERSION,
      chainId,
      revision: "1",
    },
    primaryType: PRIMARY_TYPE,
    types: {
      StarknetDomain: [
        { name: "name", type: "shortstring" },
        { name: "version", type: "shortstring" },
        { name: "chainId", type: "shortstring" },
        { name: "revision", type: "shortstring" },
      ],
      DepositorValidation: [
        { name: "depositor", type: "ContractAddress" },
        { name: "issued_at", type: "u128" },
      ],
    },
    message: { depositor, issued_at: issuedAt },
  };
}

export function signDepositorValidation(
  input: SignDepositorValidationInput
): SignDepositorValidationOutput {
  const signerPublicKey = ec.starkCurve.getStarkKey(input.signerPrivateKey);
  const issuedAt = BigInt(input.issuedAt).toString();
  const td = buildTypedData(input.depositor, issuedAt, input.chainId);
  const messageHash = typedData.getMessageHash(td, signerPublicKey);
  const signature = ec.starkCurve.sign(messageHash, input.signerPrivateKey);
  return {
    messageHash: num.toHex(messageHash),
    signature: {
      r: num.toHex(signature.r),
      s: num.toHex(signature.s),
    },
    signerPublicKey: num.toHex(signerPublicKey),
    input: {
      depositor: num.toHex(input.depositor),
      issuedAt,
      chainId: input.chainId,
    },
  };
}

function parseArgs(argv: string[]): SignDepositorValidationInput {
  const flags = new Map<string, string>();
  for (let i = 0; i < argv.length; i += 2) {
    const key = argv[i];
    const value = argv[i + 1];
    if (!key?.startsWith("--") || value === undefined) {
      throw new Error(`Invalid argument near "${key}"`);
    }
    flags.set(key.slice(2), value);
  }
  const required = ["signer-private-key", "depositor", "issued-at", "chain-id"];
  for (const name of required) {
    if (!flags.has(name)) {
      throw new Error(`Missing required --${name}`);
    }
  }
  return {
    signerPrivateKey: flags.get("signer-private-key")!,
    depositor: flags.get("depositor")!,
    issuedAt: flags.get("issued-at")!,
    chainId: flags.get("chain-id")!,
  };
}

const isCli = import.meta.url === `file://${process.argv[1]}`;
if (isCli) {
  try {
    const input = parseArgs(process.argv.slice(2));
    const result = signDepositorValidation(input);
    process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
  } catch (error) {
    process.stderr.write(`${(error as Error).message}\n`);
    process.exit(1);
  }
}

File: scripts/address_validation_signer/py/validation_signer.py
"""Reference SNIP-12 signer for the privacy contract's depositor validation.

Produces a signature that ``privacy::snip12::verify_depositor_validation``
accepts. The typed-data layout MUST stay in lockstep with the Cairo verifier
and with the sibling ``ts/validation_signer.ts`` — any drift on one side
breaks all three.

Usage:
    python validation_signer.py \\
        --signer-private-key 0x... \\
        --depositor 0x... \\
        --issued-at 1700000000 \\
        --chain-id SN_SEPOLIA

Or import ``sign_depositor_validation`` from another script/test.
"""

from __future__ import annotations

import argparse
import json
import sys
from dataclasses import dataclass

from starknet_py.hash.utils import message_signature, private_to_stark_key
from starknet_py.utils.typed_data import TypedData

SNIP12_DOMAIN_NAME = "Screening"
SNIP12_DOMAIN_VERSION = "2"
PRIMARY_TYPE = "DepositorValidation"


@dataclass(frozen=True)
class SignInput:
    signer_private_key: int
    depositor: int
    issued_at: int
    chain_id: str


def _build_typed_data(depositor: int, issued_at: int, chain_id: str) -> TypedData:
    typed_data_dict = {
        "domain": {
            "name": SNIP12_DOMAIN_NAME,
            "version": SNIP12_DOMAIN_VERSION,
            "chainId": chain_id,
            "revision": "1",
        },
        "primaryType": PRIMARY_TYPE,
        "types": {
            "StarknetDomain": [
                {"name": "name", "type": "shortstring"},
                {"name": "version", "type": "shortstring"},
                {"name": "chainId", "type": "shortstring"},
                {"name": "revision", "type": "shortstring"},
            ],
            "DepositorValidation": [
                {"name": "depositor", "type": "ContractAddress"},
                {"name": "issued_at", "type": "u128"},
            ],
        },
        "message": {
            "depositor": hex(depositor),
            "issued_at": str(issued_at),
        },
    }
    return TypedData.from_dict(typed_data_dict)


def sign_depositor_validation(sign_input: SignInput) -> dict:
    signer_public_key = private_to_stark_key(sign_input.signer_private_key)
    typed_data = _build_typed_data(
        sign_input.depositor, sign_input.issued_at, sign_input.chain_id
    )
    message_hash = typed_data.message_hash(signer_public_key)
    # seed=None selects plain RFC 6979 nonce derivation (no extra entropy).
    # starknet-py's default (seed=32) folds the seed into the nonce as extra
    # entropy, which yields a different — equally valid — (r, s) than
    # starknet.js/@scure. Plain RFC 6979 keeps this signer bit-compatible with
    # the TypeScript signers, so identical inputs produce identical signatures.
    signature_r, signature_s = message_signature(
        message_hash, sign_input.signer_private_key, seed=None
    )
    return {
        "messageHash": hex(message_hash),
        "signature": {"r": hex(signature_r), "s": hex(signature_s)},
        "signerPublicKey": hex(signer_public_key),
        "input": {
            "depositor": hex(sign_input.depositor),
            "issuedAt": str(sign_input.issued_at),
            "chainId": sign_input.chain_id,
        },
    }


def _parse_int(value: str) -> int:
    return int(value, 16) if value.lower().startswith("0x") else int(value)


def _parse_args(argv: list[str]) -> SignInput:
    parser = argparse.ArgumentParser(description="Sign a DepositorValidation under SNIP-12.")
    parser.add_argument("--signer-private-key", required=True)
    parser.add_argument("--depositor", required=True)
    parser.add_argument("--issued-at", required=True)
    parser.add_argument("--chain-id", required=True)
    args = parser.parse_args(argv)
    return SignInput(
        signer_private_key=_parse_int(args.signer_private_key),
        depositor=_parse_int(args.depositor),
        issued_at=_parse_int(args.issued_at),
        chain_id=args.chain_id,
    )


def main(argv: list[str]) -> int:
    sign_input = _parse_args(argv)
    result = sign_depositor_validation(sign_input)
    print(json.dumps(result, indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
